content/blog/2024-01-11-xilem-backend-roadmap.md

```diff
@@ -0,0 +1,157 @@
+
+### Switching to Masonry as a backend
+
+Okay, this one makes me blush a little.
+
+I'm the only maintainer of Masonry, a GUI crate with barely over 300 stars on Github and little outside interest. Can I really justify taking the money Google pays me to improve Xilem and spending my time on Masonry instead?
+
+That choice isn't random, mind you: Masonry was forked from Druid, Xilem's ancestor, and in fact Xilem and Masonry share a lot of code right now.
+
+In fact, I'd argue that Masonry's codebase is of better quality than Xilem's current native backend. This isn't meant to be a put-down of the work of Xilem maintainers; rather, it's the outcome of a difference in priorities. Nobody was very interested in Xilem's native backend (though Raph was aware of its importance, hence me getting paid to write this), and lots of people were very interested in the frontend and the web backends, therefore the frontend and web backends saw the most work. But on an objective level, I think Masonry's backend codebase is a healthier starting point.
+
+Masonry also comes with some built-in perks, like powerful unit tests and a structured widget graph.
+
+It's not clear whether we want to import the Masonry codebase, the crate, or the entire repository, and it's not clear what the ultimate branding should be. I'm writing a full RFC to lay out the different possibilities to the community, so we can reach a consensus before doing more work (including possibly not using Masonry at all).
+
+If we do move on in that direction, then Masonry will need some porting work: 
+
+- Using Vello and Parley instead of Piet.
+- Using Glazier instead of druid-shell.
+- Integrating AccessKit.
+- Removing the dependency on GTK.
+
+AccessKit support in particular is table stakes, by now. The Rust ecosystem is converging on the idea that accessibility (or at least, screen reader support) should be integrated early on in the design of GUI frameworks, not tacked on at the end.
+
+There are other things Xilem added over Druid (notably virtual lists and async support), but I believe Masonry doesn't need to implement them right away for parity. Both will be very important for eventual performance work.
+
+
+### Not using custom Widgets
+
+This is another big one, and will likely also require an RFC.
+
+We should write Xilem's backend under the assumption that end users of the library (including the Xilem frontend) won't be able to create their own widgets. Instead, they will have to compose the primitives given to them the same way they compose DOM elements in the browser. The primary escape hatch will be a fixed-size surface where end-users can use arbitrary paint primitives, like the web canvas.
+
+This is a bold move, but one I'm confident in. Developers of hobby GUI frameworks tend to provide a severely anemic set of primitives, under the assumption that if anyone wants other features in their widgets, they can implement them themselves. Both Druid and Xilem suffer from this.
```

Comment:
I think it's a bold assumption, that hobby frameworks suffer from this. I think they just suffer from the bandwidth that big GUI frameworks in production have (due to financing/man-power?). Of course there should be a big amount of high-quality primitives, such that the "escape-hatch" of writing own widgets is rarely necessary for the usual app. But especially in advanced applications (CAD, 3D-modelling, DAW, Image-editing etc.) I'm almost sure that just providing a high-level canvas like-primitive will not be enough (or rather provide bad UX when the provided widgets are not cutting it). Especially when there are built-in primitives used in such a "custom" widget.

Reply:
> But especially in advanced applications (CAD, 3D-modelling, DAW, Image-editing etc.)

This class of applications is pretty interesting. It's one I have some experience in and Some Thoughts About. Overall I think if you need to code the "rectangle you click on" part of your editor, given the choice between extending xilem's current Widget trait and a web canvas, you'd be better served starting from a canvas.

Qt gives you QGraphicsWidget for these use cases, which is basically a whole parallel class hierarchy completely distinct from regular Widgets, with the main difference being that they're manually positioned and don't do layout, which I think is the right call.

So if/when we want to support graphic editors, I think letting users extend the Widget trait won't be what we want anyway; rather, we want to create a parallel CanvasItem/GraphicsWidget/EditorWidget trait (whatever your favorite name is) with the big difference being how it handles invalidation.

Reply:
> xilem's current Widget trait and a web canvas, you'd be better served starting from a canvas

I think this really depends on the concrete use-case, a 3D view without a lot of UI in it, likely. Custom (interactable) UI elements composing different other (built-in) widgets and playing nicely together with the internal layout system and other UI components (anywhere in the tree), less likely.

> letting users extend the Widget trait won't be what we want anyway

That remains to be seen, having full control of the whole widget has its advantages. Also to avoid some kind of custom widget layer on top. I just don't see the real gain we get from adding a whole new widget layer on top of the core widget layer, only additional complexity with that.

I would like to avoid premature optimization here, and see how far we get without an additional layer (because abstractions are often limiting, and require a lot of thought to avoid limitations). A canvas widget can and will make sense non-the-less (e.g. for the mentioned 3D-view).

As mentioned here and on zulip, implementing a widget should be last resort in case full control is necessary, so I think we don't have to focus on providing best DX here (though would of course be great for internal widget dev as well) and rather on the end-user UX ("components").

As you mention Qt, Qt also allows creating custom widgets inheriting from `QWidget`...

Reply:
> Custom (interactable) UI elements composing different other (built-in) widgets and playing nicely together with the internal layout system and other UI components (anywhere in the tree), less likely.

I'm not sure we're visualizing the same thing.

But my general point is, in the kind of editor we're talking about, whether it's CAD, visual programming, circuit modelling, 3D modelling, etc, once you go into the "editing area", you have very little use for you normal widget primitives. You're going to have different interactions, no or little inline layout, selection works differently, etc.

There's *some* amount of code sharing with regular layout, but it's different enough you really want a whole separate class tree. Think "DOM nodes vs SVG nodes".

You're still going to use regular widgets in your app, but it's going to be in the gutters, in dialog files, or in `foreignObject`-ish items embedded within your graph.

> Qt also allows creating custom widgets inheriting from QWidget...

Yes, I'm aware. My point is that, for the kind of use-case you're pointing at (CAD-like editors) you wouldn't want to inherit QWidget for the editor contents anyway.

---

That said, yeah, I get the general points that there are use-cases where users want to be able to do their own stuff. I'm writing a new draft that softens the "no custom widgets" part a bit.

Reply:
Yeah current version is much better. Sorry when I came a little bit rigor and/or persistent...